sound/soc/sof/ipc4-topology.c

```diff
@@ -1462,6 +1465,15 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
 
 		msg = &pipeline->msg;
 		msg->primary |= pipeline->mem_usage;
+
+		swidget->instance_id = ida_alloc_max(&ppl_ida, ipc4_data->num_ppl, GFP_KERNEL);
+		if (swidget->instance_id < 0) {
+			dev_err(sdev->dev, "failed to assign instance id for %s: %d\n",
+				swidget->widget->name, swidget->instance_id);
+			return swidget->instance_id;
+		}
+		msg->primary &= ~SOF_IPC4_GLB_PIPE_INSTANCE_MASK;
+		msg->primary |= SOF_IPC4_GLB_PIPE_INSTANCE_ID(swidget->instance_id);
```

Comment:
I am still not clear on the concepts
pipeline_id comes from topology, it will be used to generate all the control names
instance_id is allocated with an IDA

How does this help? we still need to make sure the pipeline_id is unique in the topology, so why can't we *require* that the number of pipelines is compatible with firmware requirements.

Is this to promote topology reuse, e.g. in some cases we not include parts of the topology but we want to keep topology-defined pipeline_id fixed?

It would help to better explain the problem statement and goals.

Reply:
> I am still not clear on the concepts pipeline_id comes from topology, it will be used to generate all the control names instance_id is allocated with an IDA
> 
> How does this help? we still need to make sure the pipeline_id is unique in the topology, so why can't we _require_ that the number of pipelines is compatible with firmware requirements.

The total number of pipeline may exceed the max number of pipeline that the FW can support.
The max pipeline number is less than we thought. For example, I get the max pipeline number of TGL Up Xtreme board is 16.
And We have more than one pipelines in a PCM. If we have 2 SDW, 1 DMIC and 4 HDMI PCMs, we may need more than 16 pipelines.
The proposal is to assign instance_id in runtime. So 16 pipelines is likely enough.

> 
> Is this to promote topology reuse, e.g. in some cases we not include parts of the topology but we want to keep topology-defined pipeline_id fixed?
> 
> It would help to better explain the problem statement and goals.

Besides, assign instance_id in kernel giving more flexibility to topology.

Reply:
> > I am still not clear on the concepts pipeline_id comes from topology, it will be used to generate all the control names instance_id is allocated with an IDA
> > How does this help? we still need to make sure the pipeline_id is unique in the topology, so why can't we _require_ that the number of pipelines is compatible with firmware requirements.
> 
> The total number of pipeline may exceed the max number of pipeline that the FW can support. The max pipeline number is less than we thought. For example, I get the max pipeline number of TGL Up Xtreme board is 16. And We have more than one pipelines in a PCM. If we have 2 SDW, 1 DMIC and 4 HDMI PCMs, we may need more than 16 pipelines. The proposal is to assign instance_id in runtime. So 16 pipelines is likely enough.

I don't get it @bardliao 

Are you saying that the IDA is used when the pipeline is instantiated, so depending on how the PCMs are opened we may or may not exceed the limitations on the number of pipelines?

This would be a change from all previous topologies where the assumption was that all resources can be used concurrently. I have no idea how UCM/PipeWire/PulseAudio would behave if *sometimes* a pipeline cannot be opened. 

Actually related question: is the IDA allocation handled in open or hw_params?

Reply:
> > > I am still not clear on the concepts pipeline_id comes from topology, it will be used to generate all the control names instance_id is allocated with an IDA
> > > How does this help? we still need to make sure the pipeline_id is unique in the topology, so why can't we _require_ that the number of pipelines is compatible with firmware requirements.
> > 
> > 
> > The total number of pipeline may exceed the max number of pipeline that the FW can support. The max pipeline number is less than we thought. For example, I get the max pipeline number of TGL Up Xtreme board is 16. And We have more than one pipelines in a PCM. If we have 2 SDW, 1 DMIC and 4 HDMI PCMs, we may need more than 16 pipelines. The proposal is to assign instance_id in runtime. So 16 pipelines is likely enough.
> 
> I don't get it @bardliao
> 
> Are you saying that the IDA is used when the pipeline is instantiated, so depending on how the PCMs are opened we may or may not exceed the limitations on the number of pipelines?

Depends on how many PCMs are opened, yes.

> 
> This would be a change from all previous topologies where the assumption was that all resources can be used concurrently. I have no idea how UCM/PipeWire/PulseAudio would behave if _sometimes_ a pipeline cannot be opened.
> 

Ah, I didn't know the assumption. But what if the firmware can only support less pipelines then we need?
Say we have 1 headset, 1 speaker with echo reference, 1 DMIC, and 4 HDMI.
So we have 6 playbacks and 3 captures.
Currently we need 2 pipelines for each playback or capture. So we need 18 pipelines in the above case.
And TGL FW can only support 16 pipelines.

> Actually related question: is the IDA allocation handled in open or hw_params?

In open